components/gfx/paint_context.rs

```diff
@@ -466,6 +466,167 @@ impl<'a> PaintContext<'a> {
+                Size2D::zero()
+            } else {
+                Size2D::new(radius.width - corner_width, radius.height - corner_height)
+            }
+        }
+
+        // T = top, B = bottom, L = left, R = right
+        let ((box_TL, _),
+             (box_TR, _),
+             (box_BR, _),
+             (box_BL, _)) = PaintContext::border_bounds(bounds, border);
+
+        let origin_TL = box_TL + Point2D::new(radii.top_left.width, radii.top_left.height);
+        let origin_TR = box_TR + Point2D::new(-radii.top_right.width, radii.top_right.height);
+        let origin_BR = box_BR + Point2D::new(-radii.bottom_right.width, -radii.bottom_right.height);
+        let origin_BL = box_BL + Point2D::new(radii.bottom_left.width, -radii.bottom_left.height);
+
+        let dist_to_elbow_TL = distance_to_elbow(&radii.top_left, border.left, border.top);
+        let dist_to_elbow_TR = distance_to_elbow(&radii.top_right, border.right, border.top);
+        let dist_to_elbow_BR = distance_to_elbow(&radii.bottom_right, border.right, border.bottom);
+        let dist_to_elbow_BL = distance_to_elbow(&radii.bottom_left, border.left, border.bottom);
+
+        ((origin_TL, dist_to_elbow_TL),
+         (origin_TR, dist_to_elbow_TR),
+         (origin_BR, dist_to_elbow_BR),
+         (origin_BL, dist_to_elbow_BL))
+    }
+
+    #[allow(non_snake_case)]
+    fn draw_corner(path_builder: &mut PathBuilder,
+                   origin: &Point2D<AzFloat>,
+                   radius: &Size2D<AzFloat>,
+                   inner_border: &Point2D<AzFloat>,
+                   outer_border: &Point2D<AzFloat>,
+                   dist_elbow: &Size2D<AzFloat>,
+                   clockwise: bool) {
+        let rad_R: AzFloat = 0.;
+        let rad_BR = rad_R  + f32::consts::FRAC_PI_4;
+        let rad_B  = rad_BR + f32::consts::FRAC_PI_4;
+        let rad_BL = rad_B  + f32::consts::FRAC_PI_4;
+        let rad_L  = rad_BL + f32::consts::FRAC_PI_4;
+        let rad_TL = rad_L  + f32::consts::FRAC_PI_4;
+        let rad_T  = rad_TL + f32::consts::FRAC_PI_4;
+
+        fn compatible_borders_corner(border_corner_radius: &Size2D<f32>,
+                                     border1_width: f32,
+                                     border2_width: f32) -> bool {
+            (border_corner_radius.width - border_corner_radius.height).abs() <= f32::EPSILON &&
+                (border1_width - border2_width).abs() <= f32::EPSILON
+        }
+
+        if PaintContext::is_zero_radius(radius) {
+            return;
+        }
+        let ellipse = Ellipse { origin: *origin, width: radius.width, height: radius.height };
+        let simple_border = compatible_borders_corner(&radius,
+                                                      (outer_border.x - inner_border.x).abs(),
+                                                      (outer_border.y - inner_border.y).abs());
+        let corner_angle = match (simple_border, inner_border.x >= outer_border.x) {
```

Comment:
I like the use of pattern matching, but this seems simpler as an if and then another nested if.